drivers/gpio/gpio_nrfx.c

```diff
@@ -192,40 +68,96 @@ static int gpio_nrfx_config(const struct device *port,
 		break;
 
 	default:
-		return -EINVAL;
+		drive = NRF_GPIO_PIN_D0H1;
+		break;
+	}
+
+	return drive;
+}
+
+static nrf_gpio_pin_pull_t get_pull(gpio_flags_t flags)
+{
+	if (flags & GPIO_PULL_UP) {
+		return NRF_GPIO_PIN_PULLUP;
+	} else if (flags & GPIO_PULL_DOWN) {
+		return NRF_GPIO_PIN_PULLDOWN;
 	}
 
-	if ((flags & GPIO_PULL_UP) != 0) {
-		pull = NRF_GPIO_PIN_PULLUP;
-	} else if ((flags & GPIO_PULL_DOWN) != 0) {
-		pull = NRF_GPIO_PIN_PULLDOWN;
-	} else {
-		pull = NRF_GPIO_PIN_NOPULL;
+	return NRF_GPIO_PIN_NOPULL;
+}
+
+static int pin_uninit(nrfx_gpiote_pin_t pin)
+{
+	uint8_t ch;
+	nrfx_err_t err;
+
+	err = nrfx_gpiote_pin_uninit(pin);
+	if (err != NRFX_SUCCESS) {
+		return -EIO;
+	}
+
+	nrf_gpio_cfg_default(pin);
+	err = nrfx_gpiote_channel_get(pin, &ch);
+	if (err == NRFX_SUCCESS) {
+		err = nrfx_gpiote_channel_free(ch);
+	} else if (err == NRFX_ERROR_INVALID_PARAM) {
+		err = NRFX_SUCCESS;
+	}
+
+	return (err != NRFX_SUCCESS) ? -EIO : 0;
+}
+
+static int gpio_nrfx_pin_configure(const struct device *port, gpio_pin_t pin,
+				   gpio_flags_t flags)
+{
+	nrfx_err_t err;
+	nrfx_gpiote_pin_t abs_pin =
+		NRF_GPIO_PIN_MAP(get_port_cfg(port)->port_num, pin);
+
+	if (flags == GPIO_DISCONNECTED) {
+		return pin_uninit(abs_pin);
 	}
 
-	dir = ((flags & GPIO_OUTPUT) != 0)
-	      ? NRF_GPIO_PIN_DIR_OUTPUT
-	      : NRF_GPIO_PIN_DIR_INPUT;
+	nrfx_gpiote_trigger_config_t trigger_config = {
+		.trigger = NRFX_GPIOTE_TRIGGER_NONE
+	};
 
-	input = ((flags & GPIO_INPUT) != 0)
-		? NRF_GPIO_PIN_INPUT_CONNECT
-		: NRF_GPIO_PIN_INPUT_DISCONNECT;
+	/* Remove previously configured trigger when pin is reconfigured. */
+	err = nrfx_gpiote_input_configure(abs_pin, NULL, &trigger_config, NULL);
+	if (err != NRFX_SUCCESS) {
+		return -EINVAL;
+	}
 
-	if ((flags & GPIO_OUTPUT) != 0) {
-		if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0) {
-			nrf_gpio_port_out_set(reg, BIT(pin));
-		} else if ((flags & GPIO_OUTPUT_INIT_LOW) != 0) {
-			nrf_gpio_port_out_clear(reg, BIT(pin));
+	if (flags & GPIO_OUTPUT) {
+		nrfx_gpiote_output_config_t output_config = {
+			.drive = get_drive(flags),
+			.input_connect = (flags & GPIO_INPUT) ?
+				NRF_GPIO_PIN_INPUT_CONNECT :
+				NRF_GPIO_PIN_INPUT_DISCONNECT,
+			.pull = get_pull(flags)
+		};
+
+		if (flags & GPIO_OUTPUT_INIT_HIGH) {
+			nrf_gpio_pin_set(abs_pin);
+		} else if (flags & GPIO_OUTPUT_INIT_LOW) {
+			nrf_gpio_pin_clear(abs_pin);
```

Comment:
Since the port is already known here, you could save the calls to `nrf_gpio_pin_port_decode()` here by using `nrf_gpio_port_out_set(reg, BIT(pin))` and `nrf_gpio_port_out_clear(reg, BIT(pin))`, respectively, as it was done so far.